zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/ProcessInstanceStateTransitionGuard.java

```diff
@@ -187,4 +200,149 @@ private Either<String, ElementInstance> hasNonInterruptedFlowScope(
                   BufferUtil.bufferAsString(element.getId())));
     }
   }
+
+  private Either<String, ?> canActivateInclusiveGateway(
+      final BpmnElementContext context, final ExecutableFlowElement executableFlowElement) {
+    if (context.getBpmnElementType() == BpmnElementType.INCLUSIVE_GATEWAY) {
+      final var element = (ExecutableFlowNode) executableFlowElement;
+      final int numberOfIncomingSequenceFlows = element.getIncoming().size();
+      final int numberOfTakenSequenceFlows =
+          stateBehavior.getNumberOfTakenSequenceFlows(context.getFlowScopeKey(), element.getId());
+
+      // Accept if all incoming sequence flows were taken at least once
+      if (numberOfTakenSequenceFlows >= numberOfIncomingSequenceFlows) {
+        return Either.right(null);
+      }
+
+      if (stateBehavior.getElementInstance(context.getElementInstanceKey()) == null) {
+        return doesPathExist(context, element);
+      }
+    }
+
+    // Accept if the element is not inclusive gateway or continue processing after incident
+    // resolved.
+    return Either.right(null);
+  }
+
+  private Either<String, ?> doesPathExist(
+      final BpmnElementContext context, final ExecutableFlowElement executableFlowElement) {
+    final Optional<DeployedProcess> deployedProcess =
+        stateBehavior.getProcess(context.getProcessDefinitionKey(), context.getTenantId());
+
+    return deployedProcess
+        .map(DeployedProcess::getProcess)
+        .map(process -> doesPathExist(context, executableFlowElement, process))
+        .orElse(
+            Either.left(
+                String.format(
+                    "Expected to find a deployed process for process definition key '%d', but none found.",
+                    context.getProcessDefinitionKey())));
+  }
+
+  private Either<String, ?> doesPathExist(
+      final BpmnElementContext context,
+      final ExecutableFlowElement executableFlowElement,
+      final ExecutableProcess process) {
+    final var targetElementId = executableFlowElement.getId();
+    final var activateElementIds = findActivateElementInFlowScope(context, targetElementId);
+    final Optional<DirectBuffer> exists =
+        activateElementIds.stream()
+            .filter(
+                elementId -> doesPathExist(process, elementId, targetElementId, new HashSet<>()))
+            .findAny();
```

Comment:
:wrench: We should exclude paths ending on an incoming sequence flow of the gateway that was already taken. All incoming flows need to be taken only once.